### ISSUE TITLE
Let's encrypt handler never finishes

### ISSUE THREAD

**Issue by omartijn** [go-gitea/gitea#13933]
I'm having trouble getting Let's Encrypt to work. The handler is started but never finishes.

I'm running this on a raspberry pi 4b running the latest debian buster. I'm running the latest stable version of gitea, which is gitea-1.13.0-linux-arm-6.

I have attached a trace log from gitea as well as the configuration. 
[app.ini.txt](https://github.com/go-gitea/gitea/files/5675433/app.ini.txt)

[gitea trace.txt](https://github.com/go-gitea/gitea/files/5675437/gitea.trace.txt)

I have confirmed that the address is reachable from the outside (by doing a wget from a vps). It seems that let's encrypt just never sends a callback. The domain name is valid and has a working AAAA record on it.

Please let me know if there's anything else you need to debug the issue.

**Comment by techknowlogick** [comment]
The handler needs to always run so that it can renew a certificate as needed, and for non LE functions it passes it to the standard server handler. 

Has LE worked previously for you? From your logs it says the server has started, what does it look like in the browser? Are there any firewall rules blocking port 22/80 or 443?

**Comment by omartijn** [comment]
It has never worked for me from gitea. When I ran it the first time I had forgotten to create a rule in the firewall, so it must have initially failed. After fixing the firewall issue - and confirming that it works by connecting from a remote network it still does not issue a certificate.

If I connect on port 80 I get a "Found" response redirecting me to https. Https then doesn't work because there is no valid certificate.

So if I understand correctly, the fact that the handler runs does not necessarily means it is sending a request to Let's Encrypt, just that it's ready to receive responses. Is it possible that the initial failure (due to me forgetting firewall rules) caused it to remember it already tried and not sending a new response to LE?

**Comment by omartijn** [comment]
Just to confirm that the issue is not with my networking setup, I managed to obtain a working certificate with the following command:

`certbot certonly --standalone -d git.example.org`

**Comment by JamesYFC** [comment]
I'm getting the same thing.

I see

```
2021/02/10 05:42:19 ...s/graceful/server.go:55:NewServer() [I] Starting new server: tcp:0.0.0.0:3000 on PID: 18
2021/02/10 05:42:19 cmd/web.go:78:func1() [I] Running Let's Encrypt handler on 0.0.0.0:80
2021/02/10 05:42:19 ...s/graceful/server.go:55:NewServer() [I] Starting new server: tcp:0.0.0.0:80 on PID: 18
```

Then nothing for a long time.

When I try to access my domain it seems to not even give me a response to the get request.

I've tried a combination of port bindings and setting the redirect port configs, nothing seems to work.

currently, mine looks something like this:

```
docker-compose.yml:
ports:
  - "80:3080"
  - "443:3000"
  - "127.0.0.1:2222:22"
```

```
app.ini:
[server]
APP_DATA_PATH        = /data/gitea
DOMAIN               = xxx
SSH_DOMAIN           = xxx
HTTP_PORT            = 3000
ROOT_URL             = xxx
DISABLE_SSH          = false
SSH_PORT             = 22
SSH_LISTEN_PORT      = 22
LFS_START_SERVER     = true
LFS_CONTENT_PATH     = /data/git/lfs
LFS_JWT_SECRET       = xxx
OFFLINE_MODE         = false
PROTOCOL             = https
ENABLE_LETSENCRYPT   = true
LETSENCRYPT_ACCEPTTOS= true
LETSENCRYPT_DIRECTORY= https
LETSENCRYPT_EMAIL    = xxx
REDIRECT_OTHER_PORT  = true
PORT_TO_REDIRECT     = 3080
```

If it helps, I'm behind cloudflare.
I've also tried binding 80:3000, or 80:80 and not setting the extra REDIRECT vars.
If you wait for long enough, sometimes it continues past that and spams me with many errors like
TLS handshake error from xxxxx:xxxx: acme/autocert: missing certificate.

**Comment by zeripath** [comment]
You can't just say docker version. What version of Gitea are you running?

**Comment by JamesYFC** [comment]
@zeripath forgot to mention the version in this one. It's `v1.13.2`, I've edited the comment to add it in.

**Comment by zeripath** [comment]
Ok I think there might actually be a known bug in this - we recently merged an entirely new update for this indicating that I think our upstream was broken.

@techknowlogick would probably know more

**Comment by techknowlogick** [comment]
🧙  I've been summoned ✨ 

I've just verified that letsencrypt works as-is inside docker, and using the binary directly. So thankfully that means it isn't broken for everyone. There may be some difficulties with validating certs behind cloudflare, as we are listening on port "80" specifically for the HTTP-01 validation, but "443" for everything else, but cloudflare communicates only via the one, so if attempting to fetch the validation via "443" the logic isn't there. (quotes are used as they are placeholders for the ports that are used).

The changes that I recently made were to make our ACME more robust and use the same ACME handling that caddy (caddyserver.com) uses, which allows more than just HTTP-01, but also TLS ALPN too (and many others in future).

@JamesYFC are you able to try without cloudflare?

**Comment by JamesYFC** [comment]
@techknowlogick it was still happening when cloudflare was set to the grey cloud mode, where the traffic shouldn't be proxied by cloudflare but we still use the name. Maybe whatever that was broken was left over even after we'd turned it off, I'm not sure.

**Comment by techknowlogick** [comment]
@JamesYFC Are you able to try using the nightly build to see if that works for you (preferably on non-production data, or at very least backup your data first)?

**Comment by JamesYFC** [comment]
@techknowlogick My apologies, I can't test this as our gitea setup was pretty much obliterated a couple weeks ago as we hurried to get something else working set up. I did nothing extra other than following the various spread out sections in the tutorials though. Plus of course using cloudflare as the dns, trying proxy on vs off.

**Comment by zeripath** [comment]
As far as I understand this should have been fixed by #14177 and its associated PRs. 

Please reopen if there are still problems.
